---
title:    "Ocampo Lab mouse reprogramming"
subtitle: "OSKM gene expression"
author:   "Author: Jo√£o Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    theme: default
    code_folding: hide
editor_options:
  chunk_output_type: console
---

# Load libraries
```{r}

library(tidyverse)
library(SummarizedExperiment)
library(patchwork)

```

# Load objects
```{r}

# SummarizedExperiment object
load(file = "output/database/SummarizedExperiment_object.RData")

# DGE results
load(file = "output/database/DGE_results.RData")

```

# Load functions
```{r}

# Theme publication
source(file = "analysis/functions/theme_publication.R")

```

# Expression of the OSKM genes
```{r}

# Ensembl IDs of the OSKM genes
OSKM_gene_ids <- 
  rowData(SummarizedExperiment_object) %>% 
  as.data.frame() %>% 
  filter(gene_name %in% c("Pou5f1", "Sox2", "Klf4", "Myc")) %>% 
  pull(gene_id)

# Barplot with the OSKM gene expression per Tissue and Genotype
barplot_OSKM_expression <- 
  assay(SummarizedExperiment_object, "log2_norm_counts") %>% 
  .[OSKM_gene_ids,] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  gather(`Sample ID`, log2_norm_counts, -gene_id) %>% 
  left_join(as.data.frame(rowData(SummarizedExperiment_object))) %>% 
  dplyr::select(`Sample ID`, log2_norm_counts, gene_name) %>% 
  left_join(as.data.frame(colData(SummarizedExperiment_object)) %>% rownames_to_column("Sample ID")) %>% 
  group_by(Genotype, Tissue, gene_name, Color) %>% 
  summarise(mean_log2_norm_counts = mean(log2_norm_counts),
            sd_log2_norm_counts   = sd(log2_norm_counts)) %>% 
  ungroup %>% 
  ggplot(aes(Genotype, mean_log2_norm_counts, fill = Color)) + 
  geom_hline(yintercept = 0) +
  geom_col() +
  geom_errorbar(mapping = aes(ymin = mean_log2_norm_counts-sd_log2_norm_counts,
                              ymax = mean_log2_norm_counts+sd_log2_norm_counts), width = 0.2) +
  facet_grid(Tissue ~ gene_name) + 
  scale_fill_identity(aes(Color)) + 
  theme_publication() +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(), 
        strip.placement = "ouside", 
        strip.text.y = element_text(angle = 0, hjust = 0)) +
  labs(y = "Log2(norm. counts + 1)") +
  scale_y_continuous(expand = c(0,0))

# Heatmap with the OSKM gene expression per Tissue and Genotype
heatmap_OSKM_expression <- 
  assay(SummarizedExperiment_object, "log2_norm_counts") %>% 
  .[OSKM_gene_ids,] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  gather(`Sample ID`, log2_norm_counts, -gene_id) %>% 
  left_join(as.data.frame(rowData(SummarizedExperiment_object))) %>% 
  dplyr::select(`Sample ID`, log2_norm_counts, gene_name) %>% 
  left_join(as.data.frame(colData(SummarizedExperiment_object)) %>% rownames_to_column("Sample ID")) %>% 
  group_by(Genotype, Tissue, gene_name, Color) %>% 
  summarise(mean_log2_norm_counts = mean(log2_norm_counts),
            sd_log2_norm_counts   = sd(log2_norm_counts)) %>% 
  ungroup %>% 
  ggplot(aes(Genotype, fct_rev(Tissue), fill = mean_log2_norm_counts)) + 
  geom_tile(color = "white") +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        panel.border     = element_blank(), 
        axis.title.y     = element_blank(), 
        plot.title       = element_text(face = "bold"),
        legend.key.width = unit(0.35, "in")) +
  facet_grid( ~ gene_name) +
  labs(fill = "Log2(norm. counts + 1)", title = "OSKM RNA-seq normalized counts", subtitle = "The results presented are the mean normalized counts of the biological replicates") +
  paletteer::scale_fill_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1)

# Heatmap with the OSKM gene expression per Tissue and Genotype (scaled by row for each gene)
heatmap_OSKM_expression_scaled <- 
  assay(SummarizedExperiment_object, "log2_norm_counts") %>% 
  .[OSKM_gene_ids,] %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  gather(`Sample ID`, log2_norm_counts, -gene_id) %>% 
  left_join(as.data.frame(rowData(SummarizedExperiment_object))) %>% 
  dplyr::select(`Sample ID`, log2_norm_counts, gene_name) %>% 
  left_join(as.data.frame(colData(SummarizedExperiment_object)) %>% rownames_to_column("Sample ID")) %>% 
  group_by(Genotype, Tissue, gene_name, Color) %>% 
  summarise(mean_log2_norm_counts = mean(log2_norm_counts),
            sd_log2_norm_counts   = sd(log2_norm_counts)) %>% 
  ungroup %>% 
  group_by(Tissue, gene_name) %>% 
  mutate(scaled_mean_log2_norm_counts = scale(mean_log2_norm_counts)) %>% 
  ungroup %>% 
  ggplot(aes(Genotype, fct_rev(Tissue), fill = scaled_mean_log2_norm_counts)) + 
  geom_tile(color = "white") +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        panel.border     = element_blank(), 
        axis.title.y     = element_blank(), 
        plot.title       = element_text(face = "bold"),
        legend.key.width = unit(0.35, "in")) +
  facet_grid( ~ gene_name) +
  labs(fill = "Scaled log2(norm. counts + 1)", title = "OSKM RNA-seq normalized counts", subtitle = "The normalized counts were scaled by row for each gene") +
  paletteer::scale_fill_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1)

```

# Differential expression results for the OSKM genes by genotype and tissue
```{r}

# Dataframe with the OSKM genes DGE results by genotype and tissue
DGE_genotype_WT_tissue_OSKM <- 
  map2_dfr(DGE_results$DGE_genotype_WT_tissue, names(DGE_results$DGE_genotype_WT_tissue), function(x, y){
    
    DGE_sign_results <- 
        x %>% 
        as.data.frame() %>% 
        filter(padj < 0.05 & abs(log2FoldChange) > 2) %>% 
        rownames_to_column("gene_id") %>% 
        dplyr::select(gene_id, log2FoldChange, padj)
    
    rowData(SummarizedExperiment_object) %>% 
      as.data.frame() %>% 
      filter(gene_name %in% c("Pou5f1", "Sox2", "Klf4", "Myc")) %>%
      dplyr::select(gene_id, gene_name) %>% 
      left_join(DGE_sign_results) %>% 
      mutate(comparison = y, 
             Tissue     = str_remove(y, ".*vs.WT_") %>% str_replace_all("_", " "), 
             Genotype   = str_remove(y, ".vs.*"))
  })

# Plot with the DGE results for the OSKM genes by genotype and tissue
DGE_OSKM_plot <- 
  DGE_genotype_WT_tissue_OSKM %>% 
    mutate(Genotype = factor(Genotype, levels = levels(SummarizedExperiment_object$Genotype)),
           Tissue   = factor(Tissue, levels = levels(SummarizedExperiment_object$Tissue))) %>%
    mutate(padj = -log10(padj)) %>% 
    ggplot(aes(Genotype, fct_rev(Tissue))) + 
    geom_tile(color = "black", fill = NA) + 
    geom_point(mapping = aes(fill = log2FoldChange, size = padj), color = "black", shape = 21) + 
    facet_grid( ~ gene_name) +
    theme_publication() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), 
          panel.border     = element_blank(), 
          axis.title.y     = element_blank(),
          plot.title       = element_text(face = "bold"),
          legend.key.width = unit(0.35, "in")) + 
    scale_size_area(max_size = 15) +
    scale_fill_distiller(palette = "Reds") +
    labs(fill     = "Log2(Fold change)", 
         size     = "-Log10(Adj. P-value)",
         title    = "Differential expression results between genotypes and wild type (WT)", 
         subtitle = "Negative Log2 Fold change means the WT samples have lower expression") 

```

# Save plots
```{r}

svg(filename = "output/figures/plot1.svg", width = 11, height = 15)
patchwork::wrap_plots(barplot_OSKM_expression, heatmap_OSKM_expression, ncol = 1, byrow = 2, heights = c(1, 0.25))
dev.off()

DGE_OSKM_plot

```

# Session information
```{r Session info, echo = FALSE, results = "markup"}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.
