---
title:    "Ocampo Lab mouse reprogramming"
subtitle: "DGE results - Comparison between genotype and WT by tissue"
author:   "Author: Jo√£o Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    theme: default
    code_folding: hide
editor_options:
  chunk_output_type: console
---

# Load libraries
```{r}

library(tidyverse)
library(SummarizedExperiment)
library(patchwork)
library(UpSetR)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Mm.eg.db)

```

# Load objects
```{r}

# SummarizedExperiment object
load(file = "output/database/SummarizedExperiment_object.RData")

# DGE results
load(file = "output/database/DGE_results.RData")

```

# Load functions
```{r}

# Theme publication
source(file = "analysis/functions/theme_publication.R")

```

# Differental expressed genes between genotype and WT
```{r}

# Results from the DGE genotype vs. WT
DGE_genotype_WT_results <- 
map2(DGE_results$DGE_genotype_WT, names(DGE_results$DGE_genotype_WT), function(x, y){
  
  DGE_results_filtered <- 
    x %>% 
    as.data.frame() %>% 
    filter(padj < 0.05 & abs(log2FoldChange) > 2) %>% 
    mutate(log2FoldChange_direction = case_when(log2FoldChange < 0 ~ "negative",
                                                log2FoldChange > 0 ~ "positive",
                                                TRUE ~ NA_character_))
  
  gene_id_upregulated <- 
    DGE_results_filtered %>% 
    filter(log2FoldChange_direction == "negative") %>% 
    rownames() %>% 
    unique
  
  gene_id_downregulated <- 
    DGE_results_filtered %>% 
    filter(log2FoldChange_direction == "positive") %>% 
    rownames() %>% 
    unique
  
  up_down_regulated_table <- 
    tibble("Upregulated"   = length(gene_id_upregulated),
           "Downregulated" = length(gene_id_downregulated),
           Genotype        = str_remove(y, ".vs.WT"),
           Control         = str_remove(y, ".*.vs."))
  
  gene_ids_up_down_regulated <- 
    DGE_results_filtered %>% 
    rownames_to_column("gene_id") %>% 
    dplyr::select(gene_id, log2FoldChange_direction) %>% 
    mutate(Condition = str_remove(y, ".vs.WT")) %>% 
    mutate(log2FoldChange_direction = case_when(log2FoldChange_direction == "negative" ~ "Upregulated",
                                                log2FoldChange_direction == "positive" ~ "Downregulated",
                                                TRUE ~ NA_character_
                                                ))
  
  return(
  list("gene_id_upregulated"        = gene_id_upregulated,
       "gene_id_downregulated"      = gene_id_downregulated,
       "up_down_regulated_table"    = up_down_regulated_table,
       "gene_ids_up_down_regulated" = gene_ids_up_down_regulated)
  )
})

# Dataframe with the number of differentially expressed genes for each condition
DGE_number_df <- 
  map_dfr(DGE_genotype_WT_results, ~ .x$up_down_regulated_table) %>% 
  gather(Direction, `Number of genes`, -Genotype, -Control) %>% 
  mutate(Direction = factor(Direction, levels = rev(c("Upregulated", "Downregulated")))) %>% 
  mutate(Genotype = factor(Genotype, levels = levels(SummarizedExperiment_object$Genotype)))

# Plot with the number of differentially expressed genes for each condition
plot_DGE_number <- 
  ggplot(data = DGE_number_df, aes(x = Genotype, y = `Number of genes`, fill = Direction)) + 
    geom_col() + 
    geom_text(data = 
                DGE_number_df %>% 
                group_by(Genotype, Control) %>% 
                summarise(`Total number of genes` = sum(`Number of genes`)) %>% 
                ungroup, 
              mapping     = aes(x = Genotype, y = `Total number of genes` + 50, label = `Total number of genes`), 
              size        = 3.5, 
              inherit.aes = F) +
    scale_fill_manual(values = c("Upregulated"   = "#4292C6",
                                 "Downregulated" = "#EF3B2C")) +
    theme_publication() +
    theme(panel.grid.major.x = element_blank(),
          panel.grid.minor   = element_blank(),
          panel.border       = element_blank(), 
          strip.placement    = "ouside",
          plot.title         = element_text(face = "bold"),
          legend.title       = element_blank(),
          strip.text.y       = element_text(angle = 0, hjust = 0)) +
    scale_y_continuous(expand = c(0,0), limits = c(0, 1100)) +
    labs(title = "Number of differentially expressed genes\nbetween genotype and wild type (WT)")

# Dataframe with the percentage of differentially expressed genes for each condition
DGE_pct_df <-
  DGE_number_df %>%
  group_by(Genotype, Control) %>% 
  mutate(`Gene percentage (%)` = `Number of genes`*100/sum(`Number of genes`)) %>%
  ungroup

# Plot with the percentage of differentially expressed genes for each condition
plot_DGE_pct <- 
  ggplot(data = DGE_pct_df, aes(x = Genotype, y = `Gene percentage (%)`, fill = Direction)) + 
  geom_col() + 
  geom_text(data = 
              DGE_pct_df %>% 
              filter(Direction == "Upregulated"), 
            mapping = aes(x     = Genotype, 
                          y     = `Gene percentage (%)`- 10, 
                          label = paste0(`Number of genes`, "\n", "(", round(`Gene percentage (%)`), "%)")),
            size    = 3.5, 
            color   = "black") +
  geom_text(data = 
              DGE_pct_df %>% 
              filter(Direction == "Downregulated"), 
            mapping = aes(x     = Genotype, 
                          y     = 100-`Gene percentage (%)`+ 10, 
                          label = paste0(`Number of genes`, "\n", "(", round(`Gene percentage (%)`), "%)")),
            size  = 3.5, 
            color = "white") +
  scale_fill_manual(values = c("Upregulated"   = "#4292C6",
                               "Downregulated" = "#EF3B2C")) +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border     = element_blank(), 
        strip.placement  = "ouside", 
        plot.title       = element_text(face = "bold"),
        legend.title     = element_blank(),
        strip.text.y     = element_text(angle = 0, hjust = 0)) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  labs(title = "Percentage of differentially expressed genes\nbetween genotype and wild type (WT)")

```

# Save plots
```{r}

svg(filename = "output/figures/DGE_genotype_WT/plot_DGE_number_pct.svg", width = 5, height = 10)
plot_DGE_number / plot_DGE_pct
dev.off()

```

# Genotypes vs. WT comparison
```{r}

#################################
# Shared DEGs
#################################

# Number of differentially expressed genes between genotypes
shared_DGE <- 
  rbind(
    DGE_genotype_WT_results$`4Fj.vs.WT`$gene_ids_up_down_regulated,
    DGE_genotype_WT_results$`4FsB.vs.WT`$gene_ids_up_down_regulated,
    DGE_genotype_WT_results$`4Fk.vs.WT`$gene_ids_up_down_regulated,
    DGE_genotype_WT_results$`4FsA.vs.WT`$gene_ids_up_down_regulated
    ) %>% 
    dplyr::select(gene_id, Condition) %>% 
    spread(Condition, -gene_id) %>% 
    mutate(`4Fj`  = case_when(is.na(`4Fj`)  ~ 0, TRUE ~ 1),
           `4FsB` = case_when(is.na(`4FsB`) ~ 0, TRUE ~ 1),
           `4Fk`  = case_when(is.na(`4Fk`)  ~ 0, TRUE ~ 1),
           `4FsA` = case_when(is.na(`4FsA`) ~ 0, TRUE ~ 1)) %>% 
    UpSetR::upset(nsets = 7)

pdf(file = "output/figures/DGE_genotype_WT/shared_DGE_4Fj.vs.WT.pdf", width = 10, height = 4)
shared_DGE
dev.off()


#################################
# Median expression of DEGs
#################################

# Median expression of the differentially expressed genes
median_expression_table <- 
  map2_dfr(DGE_genotype_WT_results, names(DGE_genotype_WT_results),
          
    ~ assay(SummarizedExperiment_object, "log2_norm_counts") %>% 
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>% 
      mutate(Direction = case_when(gene_id %in% .x$gene_id_upregulated   ~ "Upregulated",
                                   gene_id %in% .x$gene_id_downregulated ~ "Downregulated",
                                   TRUE                                  ~ NA_character_)) %>%
      filter(gene_id %in% unique(c(.x$gene_id_upregulated, .x$gene_id_downregulated))) %>% 
      gather(`Sample ID`, log2_norm_counts, -gene_id, -Direction) %>% 
      left_join(as.data.frame(rowData(SummarizedExperiment_object))) %>% 
      dplyr::select(`Sample ID`, log2_norm_counts, gene_id, Direction) %>% 
      left_join(as.data.frame(colData(SummarizedExperiment_object)) %>% rownames_to_column("Sample ID")) %>% 
      group_by(Genotype,, Direction) %>%
      summarise(median_log2_norm_counts = median(log2_norm_counts)) %>%
      ungroup %>% 
      mutate(DGE_comparison = str_replace(.y, ".vs.", " vs. "))
  ) %>% 
    mutate(DGE_comparison = factor(DGE_comparison, levels = c("4Fj vs. WT",
                                                              "4FsB vs. WT",
                                                              "4Fk vs. WT",
                                                              "4FsA vs. WT")))
  
# Plot of median expression of the differentially expressed genes
median_expression_plot <-
ggplot(data    = median_expression_table,
       mapping = aes(Genotype, fct_rev(DGE_comparison), fill = median_log2_norm_counts)) + 
  geom_tile(color = "white") +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        panel.border     = element_blank(), 
        axis.title.y     = element_blank(), 
        plot.title       = element_text(face = "bold"),
        legend.key.width = unit(0.35, "in")) +
  facet_grid(Direction ~ .) +
  labs(fill = "Log2(norm. counts + 1)", title = "Expression of the DEGs\nbetween genotypes and WT", subtitle = "The results presented are the median normalized\ncounts of the biological replicates and genes") +
  paletteer::scale_fill_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0))

pdf(file = "output/figures/DGE_genotype_WT/median_expression_plot_4Fj.vs.WT.pdf", width = 4.5, height = 7)
median_expression_plot
dev.off()


#################################
# enrichGO
#################################

# enrichGO upregulated
enrichGO_clusters_upregulated <-
  compareCluster(geneClusters = 
    list("4Fj upregulated"  = DGE_genotype_WT_results$`4Fj.vs.WT`$gene_id_upregulated,
         "4FsB upregulated" = DGE_genotype_WT_results$`4FsB.vs.WT`$gene_id_upregulated,
         "4Fk upregulated"  = DGE_genotype_WT_results$`4Fk.vs.WT`$gene_id_upregulated,
         "4FsA upregulated" = DGE_genotype_WT_results$`4FsA.vs.WT`$gene_id_upregulated
         ), 
         fun           = "enrichGO", 
         OrgDb         = "org.Mm.eg.db",
         keyType       = 'ENSEMBL',
         ont           = "BP",
         pAdjustMethod = "BH",
         qvalueCutoff  = 0.05,
         pvalueCutoff  = 0.05,
         readable      = TRUE)

# enrichGO downregulated
enrichGO_clusters_downregulated <-
  compareCluster(geneClusters = 
    list("4Fj downregulated"  = DGE_genotype_WT_results$`4Fj.vs.WT`$gene_id_downregulated,
         "4FsB downregulated" = DGE_genotype_WT_results$`4FsB.vs.WT`$gene_id_downregulated,
         "4Fk downregulated"  = DGE_genotype_WT_results$`4Fk.vs.WT`$gene_id_downregulated,
         "4FsA downregulated" = DGE_genotype_WT_results$`4FsA.vs.WT`$gene_id_downregulated
         ), 
         fun           = "enrichGO", 
         OrgDb         = "org.Mm.eg.db",
         keyType       = 'ENSEMBL',
         ont           = "BP",
         pAdjustMethod = "BH",
         qvalueCutoff  = 0.05,
         pvalueCutoff  = 0.05,
         readable      = TRUE)

# enrichGO upregulated dotplot
enrichGO_clusters_upregulated_dotplot <- 
clusterProfiler::dotplot(enrichGO_clusters_upregulated, group = F, showCategory = 5) +
  scale_color_distiller(palette = "Reds", direction = 1) +
  scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 50), drop = FALSE) +
  labs(fill = "Adj. p-value", size = "Gene ratio") +
  theme(panel.grid.major = element_line(linetype = 2), 
        axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
        axis.title.x     = element_blank()) +
  scale_fill_distiller(palette = "Reds", direction = -1)

# enrichGO downregulated dotplot
enrichGO_clusters_downregulated_dotplot <- 
clusterProfiler::dotplot(enrichGO_clusters_downregulated, group = F, showCategory = 5) +
  scale_color_distiller(palette = "Reds", direction = 1) +
  scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 50), drop = FALSE) +
  labs(fill = "Adj. p-value", size = "Gene ratio") +
  theme(panel.grid.major = element_line(linetype = 2), 
        axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
        axis.title.x     = element_blank()) +
  scale_fill_distiller(palette = "Reds", direction = -1)

# Combined enrichGO dotplots
pdf(file = "output/figures/DGE_genotype_WT/enrichGO_clusters_dotplot_4Fj.vs.WT.pdf", width = 15, height = 7)
enrichGO_clusters_upregulated_dotplot + enrichGO_clusters_downregulated_dotplot
dev.off()

# enrichGO upregulated emapplot
pdf(file = "output/figures/DGE_genotype_WT/emapplot_upregulated_4Fj.vs.WT.pdf", width = 7, height = 7)
emapplot(enrichplot::pairwise_termsim(enrichGO_clusters_upregulated), showCategory = 5) + 
    scale_fill_manual(values = c("#2F8CCD", "#FF9000", "#EB005C", "#00943D") %>% 
                        set_names(paste(levels(SummarizedExperiment_object$Genotype)[2:5], "upregulated")))
dev.off()

# enrichGO downregulated emapplot
pdf(file = "output/figures/DGE_genotype_WT/emapplot_downregulated_4Fj.vs.WT.pdf", width = 7, height = 7)
emapplot(enrichplot::pairwise_termsim(enrichGO_clusters_downregulated), showCategory = 5) + 
    scale_fill_manual(values = c("#2F8CCD", "#FF9000", "#EB005C", "#00943D") %>% 
                        set_names(paste(levels(SummarizedExperiment_object$Genotype)[2:5], "downregulated")))
dev.off()


#################################
# enrichKEGG
#################################

ensembl_id_to_entrez_id <- 
  function(gene_id_vector){
    mapIds(x         = org.Mm.eg.db, 
           keys      = gene_id_vector,
           column    = "ENTREZID",
           keytype   = "ENSEMBL",
           multiVals = "first") %>% 
      na.omit
  }

# enrichKEGG upregulated
enrichKEGG_clusters_upregulated <-
compareCluster(geneClusters = 
    list(
         "4Fj upregulated"  = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4Fj.vs.WT`$gene_id_upregulated),
         "4FsB upregulated" = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4FsB.vs.WT`$gene_id_upregulated),
         "4Fk upregulated"  = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4Fk.vs.WT`$gene_id_upregulated),
         "4FsA upregulated" = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4FsA.vs.WT`$gene_id_upregulated)
         ), 
         fun           = "enrichKEGG", 
         organism      = "mmu",
         pAdjustMethod = "BH",
         qvalueCutoff  = 0.05,
         pvalueCutoff  = 0.05)

# enrichKEGG downregulated
enrichKEGG_clusters_downregulated <-
compareCluster(geneClusters = 
    list(
         "4Fj downregulated"  = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4Fj.vs.WT`$gene_id_downregulated),
         "4FsB downregulated" = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4FsB.vs.WT`$gene_id_downregulated),
         "4Fk downregulated"  = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4Fk.vs.WT`$gene_id_downregulated),
         "4FsA downregulated" = ensembl_id_to_entrez_id(DGE_genotype_WT_results$`4FsA.vs.WT`$gene_id_downregulated)
         ), 
         fun           = "enrichKEGG", 
         organism      = "mmu",
         pAdjustMethod = "BH",
         qvalueCutoff  = 0.05,
         pvalueCutoff  = 0.05)

# enrichKEGG upregulated dotplot
enrichKEGG_clusters_upregulated_dotplot <- 
clusterProfiler::dotplot(enrichKEGG_clusters_upregulated, group = F, showCategory = 5) +
  scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 50), drop = FALSE) +
  labs(color = "Adj. p-value", size = "Gene ratio") +
  theme(panel.grid.major = element_line(linetype = 2), 
        axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
        axis.title.x     = element_blank()) +
  scale_fill_distiller(palette = "Reds", direction = -1)

# enrichKEGG downregulated in 4Fj dotplot
enrichKEGG_clusters_downregulated_dotplot <- 
clusterProfiler::dotplot(enrichKEGG_clusters_downregulated, group = F, showCategory = 5) +
  scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 50), drop = FALSE) +
  labs(color = "Adj. p-value", size = "Gene ratio") +
  theme(panel.grid.major = element_line(linetype = 2), 
        axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
        axis.title.x     = element_blank()) +
  scale_fill_distiller(palette = "Reds", direction = -1)

# Combined enrichKEGG dotplots
pdf(file = "output/figures/DGE_genotype_WT/enrichKEGG_4Fj.vs.WT.pdf", width = 9, height = 12)
enrichKEGG_clusters_upregulated_dotplot / enrichKEGG_clusters_downregulated_dotplot
dev.off()


#################################
# Volcano plot
#################################

volcano_plots <- 
map2(DGE_results$DGE_genotype_WT, names(DGE_results$DGE_genotype_WT), function(x, y){
  
  DGE_results_w_gene_anno <- 
    x %>% 
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>% 
      left_join(as.data.frame(rowData(SummarizedExperiment_object))) %>% 
      mutate(log2FoldChange = -log2FoldChange)
  
  geom_label_data <- 
    rbind(
      DGE_results_w_gene_anno %>% filter(abs(log2FoldChange) > 2 & padj < 0.05) %>% arrange(padj) %>% head(10),
      DGE_results_w_gene_anno %>% filter(abs(log2FoldChange) > 2 & padj < 0.05) %>% arrange(dplyr::desc(abs(log2FoldChange))) %>% head(15)
    ) %>% 
    unique
  
  non_signficant_data <- 
    DGE_results_w_gene_anno %>% 
    filter(abs(log2FoldChange) <= 2 | padj >= 0.05)
  
  signficant_data <- 
    DGE_results_w_gene_anno %>% 
    filter(abs(log2FoldChange) > 2 & padj < 0.05)
  
  ggplot(data = DGE_results_w_gene_anno, aes(log2FoldChange, -log10(padj), color = -log10(padj))) + 
    geom_hline(yintercept = 0) +
    geom_vline(xintercept = 0) +
    geom_hline(yintercept = -log10(0.05), color = "darkgray", lty = 2) +
    geom_vline(xintercept = 2, color = "darkgray", lty = 2) +
    geom_vline(xintercept = -2, color = "darkgray", lty = 2) +
    geom_point(data = non_signficant_data, color = "gray") +
    geom_point(data = signficant_data, size = 3) +
    ggrepel::geom_label_repel(data = geom_label_data, mapping = aes(label = gene_name), color = "black", box.padding = unit(0.1, "in")) +
    theme_publication() +
    scale_x_continuous(limits = ggpmisc::symmetric_limits) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), 
          legend.direction = "vertical", 
          legend.position  = "right", 
          plot.title       = element_text(face = "bold")) +
    labs(x     = "Log2 Fold change", 
         y     = "-Log10 adj. P-value", 
         color = "-Log10 adj. P-value", 
         title = paste0("Differential gene expression between ", str_remove(y, ".vs.WT"), " and WT"), 
         ) +
    paletteer::scale_color_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1)

})

pdf(file = "output/figures/DGE_genotype_WT/volcano_plots_4Fj.vs.WT.pdf", width = 14, height = 9)
volcano_plots$`4Fj.vs.WT` + 
  volcano_plots$`4FsB.vs.WT` + 
  volcano_plots$`4Fk.vs.WT` +
  volcano_plots$`4FsA.vs.WT` +
  patchwork::plot_layout(ncol = 2, nrow = 2)
dev.off()

```

# Session information
```{r Session info, echo = FALSE, results = "markup"}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.
