---
title:    "Ocampo Lab mouse reprogramming"
subtitle: "DGE results"
author:   "Author: Jo√£o Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    theme: default
    code_folding: hide
editor_options:
  chunk_output_type: console
---

# Load libraries
```{r}

library(tidyverse)
library(SummarizedExperiment)
library(patchwork)
library(UpSetR)

```

# Load objects
```{r}

# SummarizedExperiment object
load(file = "output/database/SummarizedExperiment_object.RData")

# DGE results
load(file = "output/database/DGE_results.RData")

```

# Load functions
```{r}

# Theme publication
source(file = "analysis/functions/theme_publication.R")

```

# Differental expressed genes between genotype and WT by tissue
```{r}
# Results from the DGE genotype vs. WT by tissue results
DGE_genotype_WT_tissue_results <- 
map2(DGE_results$DGE_genotype_WT_tissue, names(DGE_results$DGE_genotype_WT_tissue), function(x, y){
  
  DGE_results_filtered <- 
    x %>% 
    as.data.frame() %>% 
    filter(padj < 0.05 & abs(log2FoldChange) > 2) %>% 
    mutate(log2FoldChange_direction = case_when(log2FoldChange < 0 ~ "negative",
                                                log2FoldChange > 0 ~ "positive",
                                                TRUE ~ NA_character_))
  
  gene_id_upregulated <- 
    DGE_results_filtered %>% 
    filter(log2FoldChange_direction == "negative") %>% 
    rownames() %>% 
    unique
  
  gene_id_downregulated <- 
    DGE_results_filtered %>% 
    filter(log2FoldChange_direction == "positive") %>% 
    rownames() %>% 
    unique
  
  up_down_regulated_table <- 
    tibble("Upregulated"   = length(gene_id_upregulated),
           "Downregulated" = length(gene_id_downregulated),
           Tissue          = str_remove(y, ".*vs.WT_") %>% str_replace_all("_", " "),
           Genotype        = str_remove(y, ".vs.WT_.*"),
           Control         = str_remove(y, ".*.vs.") %>% str_remove("_.*"))
  
  gene_ids_up_down_regulated <- 
    DGE_results_filtered %>% 
    rownames_to_column("gene_id") %>% 
    dplyr::select(gene_id, log2FoldChange_direction) %>% 
    mutate(Condition = paste(str_remove(y, ".*.vs.WT_") %>% str_replace_all("_", " "), str_remove(y, ".vs.WT_.*"))) %>% 
    mutate(log2FoldChange_direction = case_when(log2FoldChange_direction == "negative" ~ "Upregulated",
                                                log2FoldChange_direction == "positive" ~ "Downregulated",
                                                TRUE ~ NA_character_
                                                ))
  
  return(
  list("gene_id_upregulated"        = gene_id_upregulated,
       "gene_id_downregulated"      = gene_id_downregulated,
       "up_down_regulated_table"    = up_down_regulated_table,
       "gene_ids_up_down_regulated" = gene_ids_up_down_regulated)
  )
})

# Dataframe with the number of differentially expressed genes for each condition
DGE_number_df <- 
  map_dfr(DGE_genotype_WT_tissue_results, ~ .x$up_down_regulated_table) %>% 
  gather(Direction, `Number of genes`, -Tissue, -Genotype, -Control) %>% 
  mutate(Direction = factor(Direction, levels = rev(c("Upregulated", "Downregulated")))) %>% 
  mutate(Tissue = factor(Tissue, levels = levels(SummarizedExperiment_object$Tissue))) %>%
  mutate(Genotype = factor(Genotype, levels = levels(SummarizedExperiment_object$Genotype)))

# Plot with the number of differentially expressed genes for each condition
plot_DGE_number <- 
  ggplot(data = DGE_number_df, aes(x = Genotype, y = `Number of genes`, fill = Direction)) + 
    geom_col() + 
    geom_text(data = 
                DGE_number_df %>% 
                group_by(Tissue, Genotype, Control) %>% 
                summarise(`Total number of genes` = sum(`Number of genes`)) %>% 
                ungroup, 
              mapping     = aes(x = Genotype, y = `Total number of genes` + 175, label = `Total number of genes`), 
              size        = 3.5, 
              inherit.aes = F) +
    facet_grid(~ Tissue, scales = "free", space = "free") +
    scale_fill_manual(values = c("Upregulated"   = "#4292C6",
                                 "Downregulated" = "#EF3B2C")) +
    theme_publication() +
    theme(panel.grid.major.x = element_blank(),
          panel.grid.minor   = element_blank(),
          panel.border       = element_blank(), 
          strip.placement    = "ouside", 
          legend.title       = element_blank(),
          strip.text.y       = element_text(angle = 0, hjust = 0)) +
    scale_y_continuous(expand = c(0,0), limits = c(0, 4000))

# Dataframe with the percentage of differentially expressed genes for each condition
DGE_pct_df <-
  DGE_number_df %>%
  group_by(Tissue, Genotype, Control) %>% 
  mutate(`Gene percentage (%)` = `Number of genes`*100/sum(`Number of genes`)) %>%
  ungroup

# Plot with the percentage of differentially expressed genes for each condition
plot_DGE_pct <- 
  ggplot(data = DGE_pct_df, aes(x = Genotype, y = `Gene percentage (%)`, fill = Direction)) + 
  geom_col() + 
  geom_text(data = 
              DGE_pct_df %>% 
              filter(Direction == "Upregulated"), 
            mapping = aes(x     = Genotype, 
                          y     = `Gene percentage (%)`- 10, 
                          label = paste0(`Number of genes`, "\n", "(", round(`Gene percentage (%)`), "%)")),
            size    = 3.5, 
            color   = "black") +
  geom_text(data = 
              DGE_pct_df %>% 
              filter(Direction == "Downregulated"), 
            mapping = aes(x     = Genotype, 
                          y     = 100-`Gene percentage (%)`+ 10, 
                          label = paste0(`Number of genes`, "\n", "(", round(`Gene percentage (%)`), "%)")),
            size  = 3.5, 
            color = "white") +
  facet_grid(~ Tissue, scales = "free", space = "free") +
  scale_fill_manual(values = c("Upregulated"   = "#4292C6",
                               "Downregulated" = "#EF3B2C")) +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border     = element_blank(), 
        strip.placement  = "ouside", 
        legend.title     = element_blank(),
        strip.text.y     = element_text(angle = 0, hjust = 0)) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0))

```

# 4Fj vs. WT comparison
```{r}

# Number of differentially expressed genes shared between tissues
shared_DGE <- 
  rbind(
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Kidney`$gene_ids_up_down_regulated,
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Liver`$gene_ids_up_down_regulated,
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Small_Intestine`$gene_ids_up_down_regulated,
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Heart`$gene_ids_up_down_regulated,
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Brain`$gene_ids_up_down_regulated,
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Skeletal_muscle`$gene_ids_up_down_regulated,
    DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Spleen`$gene_ids_up_down_regulated
    ) %>% 
    dplyr::select(gene_id, Condition) %>% 
    spread(Condition, -gene_id) %>% 
    mutate(`Kidney 4Fj`          = case_when(is.na(`Kidney 4Fj`)           ~ 0, TRUE ~ 1),
           `Liver 4Fj`           = case_when(is.na(`Liver 4Fj`)            ~ 0, TRUE ~ 1),
           `Small Intestine 4Fj` = case_when(is.na(`Small Intestine 4Fj`)  ~ 0, TRUE ~ 1),
           `Heart 4Fj`           = case_when(is.na(`Heart 4Fj`)            ~ 0, TRUE ~ 1),
           `Brain 4Fj`           = case_when(is.na(`Brain 4Fj`)            ~ 0, TRUE ~ 1),
           `Skeletal muscle 4Fj` = case_when(is.na(`Skeletal muscle 4Fj`)  ~ 0, TRUE ~ 1),
           `Spleen 4Fj`          = case_when(is.na(`Spleen 4Fj`)           ~ 0, TRUE ~ 1)) %>% 
    UpSetR::upset(nsets = 7)

# Median expression of the differentially expressed genes
median_expression_table <- 
  map2_dfr(DGE_genotype_WT_tissue_results[1:7], names(DGE_genotype_WT_tissue_results[1:7]),
          
    ~ assay(SummarizedExperiment_object, "log2_norm_counts") %>% 
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>% 
      mutate(Direction = case_when(gene_id %in% .x$gene_id_upregulated   ~ "Upregulated",
                                   gene_id %in% .x$gene_id_downregulated ~ "Downregulated",
                                   TRUE                                  ~ NA_character_)) %>%
      filter(gene_id %in% unique(c(.x$gene_id_upregulated, .x$gene_id_downregulated))) %>% 
      gather(`Sample ID`, log2_norm_counts, -gene_id, -Direction) %>% 
      left_join(as.data.frame(rowData(SummarizedExperiment_object))) %>% 
      dplyr::select(`Sample ID`, log2_norm_counts, gene_id, Direction) %>% 
      left_join(as.data.frame(colData(SummarizedExperiment_object)) %>% rownames_to_column("Sample ID")) %>% 
      group_by(Genotype, Tissue, Direction) %>%
      summarise(median_log2_norm_counts = median(log2_norm_counts)) %>%
      ungroup %>% 
      mutate(Tissue_comparison = str_remove(.y, ".*.vs.WT_") %>% str_replace_all("_", " ")) %>% 
      mutate(Tissue_comparison = factor(Tissue_comparison, levels = levels(SummarizedExperiment_object$Tissue)))
    
  )

# Plot of median expression of the differentially expressed genes
median_expression_plot <-
ggplot(data    = median_expression_table,
       mapping = aes(Genotype, fct_rev(Tissue), fill = median_log2_norm_counts)) + 
  geom_tile(color = "white") +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        panel.border     = element_blank(), 
        axis.title.y     = element_blank(), 
        plot.title       = element_text(face = "bold"),
        legend.key.width = unit(0.35, "in")) +
  facet_grid(Direction ~ Tissue_comparison) +
  labs(fill = "Log2(norm. counts + 1)", title = "Expression of the DEGs between 4Fj and WT", subtitle = "The results presented are the median normalized counts of the biological replicates and genes") +
  paletteer::scale_fill_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1)

DGE_genotype_WT_tissue_results$`4Fj.vs.WT_Spleen`$gene_id_upregulated


```


# Save plots
```{r}


```

# Session information
```{r Session info, echo = FALSE, results = "markup"}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.
