---
title:    "Comparative analysis of mouse strains for in vivo reprogramming"
subtitle: "DGE results - Comparison between genotype and WT by tissue"
author:   "Author: Jo√£o Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    theme: default
    code_folding: hide
editor_options:
  chunk_output_type: console
---

# Load libraries
```{r}

library(tidyverse)
library(SummarizedExperiment)
library(patchwork)
library(UpSetR)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Mm.eg.db)

```

# Load objects
```{r}

# Log2 normalized counts object
load(file = "output/database/log2_norm_counts_object.RData")

# SummarizedExperiment object
load(file = "output/database/SummarizedExperiment_object.RData")

# DGE results
load(file = "output/database/DGE_results.RData")

# Gene annotation
load(file = "output/database/edb_gene_annotation.RData")

```

# Load functions
```{r}

# Theme publication
source(file = "analysis/functions/theme_publication.R")

```

# Differental expressed genes between genotype and WT by tissue
```{r}

# Results from the DGE genotype vs. WT by tissue results
DGE_genotype_WT_tissue_results <- 
map2(DGE_results$DGE_genotype_WT_tissue, names(DGE_results$DGE_genotype_WT_tissue), function(x, y){
  
  DGE_results_filtered <- 
    x %>% 
    as.data.frame() %>% 
    filter(padj < 0.05 & abs(log2FoldChange) > 2) %>% 
    mutate(log2FoldChange_direction = case_when(log2FoldChange < 0 ~ "negative",
                                                log2FoldChange > 0 ~ "positive",
                                                TRUE ~ NA_character_))
  
  geneID_upregulated <- 
    DGE_results_filtered %>% 
    filter(log2FoldChange_direction == "positive") %>% 
    rownames() %>% 
    unique
  
  geneID_downregulated <- 
    DGE_results_filtered %>% 
    filter(log2FoldChange_direction == "negative") %>% 
    rownames() %>% 
    unique
  
  up_downregulated_table <- 
    tibble("Upregulated"   = length(geneID_upregulated),
           "Downregulated" = length(geneID_downregulated),
           Tissue          = str_remove(y, ".*vs.WT_") %>% str_replace_all("_", " "),
           Genotype        = str_remove(y, ".vs.WT_.*"),
           Control         = str_remove(y, ".*.vs.") %>% str_remove("_.*"))
  
  geneID_up_downregulated <- 
    DGE_results_filtered %>% 
    rownames_to_column("gene_id") %>% 
    dplyr::select(gene_id, log2FoldChange_direction) %>% 
    mutate(Condition = paste(str_remove(y, ".*.vs.WT_") %>% str_replace_all("_", " "), str_remove(y, ".vs.WT_.*"))) %>% 
    mutate(log2FoldChange_direction = case_when(log2FoldChange_direction == "positive" ~ "Upregulated",
                                                log2FoldChange_direction == "negative" ~ "Downregulated",
                                                TRUE ~ NA_character_
                                                ))
  
  return(
  list("geneID_upregulated"      = geneID_upregulated,
       "geneID_downregulated"    = geneID_downregulated,
       "up_downregulated_table"  = up_downregulated_table,
       "geneID_up_downregulated" = geneID_up_downregulated)
  )
})

# Dataframe with the number of differentially expressed genes for each condition
DGE_number_df <- 
  map_dfr(DGE_genotype_WT_tissue_results, ~ .x$up_downregulated_table) %>% 
  gather(Direction, `Number of genes`, -Tissue, -Genotype, -Control) %>% 
  mutate(Direction = factor(Direction, levels = rev(c("Upregulated", "Downregulated")))) %>% 
  mutate(Tissue = factor(Tissue, levels = levels(SummarizedExperiment_object$Tissue))) %>%
  mutate(Genotype = factor(Genotype, levels = levels(SummarizedExperiment_object$Genotype)))

# Plot with the number of differentially expressed genes for each condition
plot_DGE_number <- 
  ggplot(data = DGE_number_df, aes(x = Genotype, y = `Number of genes`, fill = Direction)) + 
    geom_col() + 
    geom_text(data = 
                DGE_number_df %>% 
                group_by(Tissue, Genotype, Control) %>% 
                summarise(`Total number of genes` = sum(`Number of genes`)) %>% 
                ungroup, 
              mapping     = aes(x = Genotype, y = `Total number of genes` + 150, label = `Total number of genes`), 
              size        = 3.5, 
              inherit.aes = F) +
    facet_grid(~ Tissue, scales = "free", space = "free") +
    scale_fill_manual(values = c("Upregulated"   = "#4292C6",
                                 "Downregulated" = "#EF3B2C")) +
    theme_publication() +
    theme(panel.grid.major.x = element_blank(),
          panel.grid.minor   = element_blank(),
          panel.border       = element_blank(), 
          strip.placement    = "ouside",
          plot.title         = element_text(face = "bold"),
          legend.title       = element_blank(),
          strip.text.y       = element_text(angle = 0, hjust = 0)) +
    scale_y_continuous(expand = c(0,0), limits = c(0, 4000)) +
    labs(title = "Number of differentially expressed genes between 4F strains and wild type (WT)")

# Dataframe with the percentage of differentially expressed genes for each condition
DGE_pct_df <-
  DGE_number_df %>%
  group_by(Tissue, Genotype, Control) %>% 
  mutate(`Gene percentage (%)` = `Number of genes`*100/sum(`Number of genes`)) %>%
  ungroup

# Plot with the percentage of differentially expressed genes for each condition
plot_DGE_pct <- 
  ggplot(data = DGE_pct_df, aes(x = Genotype, y = `Gene percentage (%)`, fill = Direction)) + 
  geom_col() + 
  geom_text(data = 
              DGE_pct_df %>% 
              filter(Direction == "Upregulated"), 
            mapping = aes(x     = Genotype, 
                          y     = `Gene percentage (%)`- 10, 
                          label = paste0(`Number of genes`, "\n", "(", round(`Gene percentage (%)`), "%)")),
            size    = 3.5, 
            color   = "black") +
  geom_text(data = 
              DGE_pct_df %>% 
              filter(Direction == "Downregulated"), 
            mapping = aes(x     = Genotype, 
                          y     = 100-`Gene percentage (%)`+ 10, 
                          label = paste0(`Number of genes`, "\n", "(", round(`Gene percentage (%)`), "%)")),
            size  = 3.5, 
            color = "white") +
  facet_grid(~ Tissue, scales = "free", space = "free") +
  scale_fill_manual(values = c("Upregulated"   = "#4292C6",
                               "Downregulated" = "#EF3B2C")) +
  theme_publication() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border     = element_blank(), 
        strip.placement  = "ouside", 
        plot.title       = element_text(face = "bold"),
        legend.title     = element_blank(),
        strip.text.y     = element_text(angle = 0, hjust = 0), 
        axis.title.x     = element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  labs(title = "Percentage of differentially expressed genes between 4F strains and wild type (WT)")

```

# Shared DEGs
```{r}

# Number of differentially expressed genes shared between tissues
shared_DGE <- 
  map_dfr(DGE_genotype_WT_tissue_results, ~ rbind(.x$geneID_up_downregulated)) %>% 
  dplyr::select(gene_id, Condition) %>% 
  spread(Condition, -gene_id) %>% 
  mutate_at(.vars = vars(-c("gene_id")), .funs = ~ as.numeric(as.factor(.x)) %>% replace_na(replace = 0))

```

# Genotype vs. WT comparison by tissue
```{r}

#################################
# Median expression of DEGs
#################################

# Median expression of the differentially expressed genes
median_expression_table <- 
  map2_dfr(DGE_genotype_WT_tissue_results, names(DGE_genotype_WT_tissue_results),
          
    ~ log2_norm_counts_object$log2_norm_counts_long %>% 
      mutate(Direction = case_when(gene_id %in% .x$geneID_upregulated   ~ "Upregulated",
                                   gene_id %in% .x$geneID_downregulated ~ "Downregulated",
                                   TRUE                                 ~ NA_character_)) %>%
      filter(gene_id %in% unique(c(.x$geneID_upregulated, .x$geneID_downregulated))) %>% 
      group_by(Genotype, Tissue, Direction) %>%
      summarise(median_log2_norm_counts = median(log2_norm_counts)) %>%
      ungroup %>%
      mutate(comparison = .y) %>% 
      mutate(Tissue_comparison = str_remove(.y, ".*.vs.WT_") %>% str_replace_all("_", " ")) %>% 
      mutate(Tissue_comparison = factor(Tissue_comparison, levels = levels(log2_norm_counts_object$log2_norm_counts_long$Tissue)))
    
  )

# Plot of median expression of the differentially expressed genes
median_expression_plot <-
map(c("4Fj", '4FsB', "4Fk", "4FsA") %>% set_names(.,.), function(x){
  
  ggplot(data    = median_expression_table %>% filter(str_detect(comparison, x)),
         mapping = aes(Genotype, fct_rev(Tissue), fill = median_log2_norm_counts)) + 
    geom_tile(color = "white") +
    theme_publication() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), 
          panel.border     = element_blank(), 
          axis.title.y     = element_blank(), 
          plot.title       = element_text(face = "bold"),
          legend.key.width = unit(0.35, "in")) +
    facet_grid(Direction ~ Tissue_comparison) +
    labs(fill = "Log2(norm. counts + 1)", title = paste0("Expression of the DEGs between ", x," and WT"), subtitle = "The results presented are the median normalized counts of the biological replicates and genes") +
    paletteer::scale_fill_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1)
  
  })


#################################
# gprofiler2
#################################

future::plan("multisession", workers = 26)

gprofiler2_upregulated_plots <- 
  furrr::future_map2(DGE_genotype_WT_tissue_results, names(DGE_genotype_WT_tissue_results), function(x, y){
    
    gprofiler2_clusters <- 
      gprofiler2::gost(query                       = x$geneID_upregulated, 
                       organism                    = "mmusculus", 
                       significant                 = T, 
                       user_threshold              = 0.05, 
                       correction_method           = "bonferroni", 
                       exclude_iea                 = T, 
                       measure_underrepresentation = F, 
                       highlight                   = T, 
                       custom_bg                   = unique(na.omit(edb_gene_annotation$gene_id)))
    
    gprofiler2_plot_data <- 
      gprofiler2_clusters$result %>% 
      mutate(p_adjust = p.adjust(p_value, "bonferroni")) %>% 
      filter(p_adjust < 0.05) %>% 
      filter(term_size < 1000) %>% 
      filter(!source %in% c("TF")) %>% 
      arrange(dplyr::desc(p_adjust)) %>% 
      tail(25) %>% 
      mutate(term_name = factor(term_name, levels = unique(term_name)))
    
    gprofiler2_plot <- 
      ggplot(data    = gprofiler2_plot_data, 
             mapping = aes(x = -log10(p_adjust), y = term_name, fill = -log10(p_adjust), label = paste0(intersection_size,"/", term_size))) + 
      geom_col(show.legend = F) + 
      geom_text(mapping = aes(x = -log10(p_adjust) + .2), hjust = 0) +
      facet_grid(source ~ ., scales = "free", space = "free") + 
      theme_publication() + 
      scale_fill_distiller(palette = "Reds", direction = 1) + 
      theme(strip.text.y       = element_text(angle = 0, hjust = 0), 
            panel.grid.minor.y = element_blank(), 
            panel.grid.major.y = element_blank(), 
            axis.title.y       = element_blank()) +
      scale_x_continuous(expand = expansion(mult = c(0, .2))) +
      labs(x        = "-Log10(adjusted p-value)", 
           fill     = "-Log10(adjusted p-value)", 
           title    = "g:Profiler functional enrichment", 
           subtitle = paste0("Upregulated genes in ", 
                             str_remove(y, ".vs.WT_.*"), 
                             "\n", 
                             "Query size = ", 
                             unique(gprofiler2_plot_data$query_size), 
                             " genes"))
    
    return(gprofiler2_plot)
  })  

gprofiler2_downregulated_plots <- 
  furrr::future_map2(DGE_genotype_WT_tissue_results, names(DGE_genotype_WT_tissue_results), function(x, y){
    
    gprofiler2_clusters <- 
      gprofiler2::gost(query                       = x$geneID_downregulated, 
                       organism                    = "mmusculus", 
                       significant                 = T, 
                       user_threshold              = 0.05, 
                       correction_method           = "bonferroni", 
                       exclude_iea                 = T, 
                       measure_underrepresentation = F, 
                       highlight                   = T, 
                       custom_bg                   = unique(na.omit(edb_gene_annotation$gene_id)))
    
    gprofiler2_plot_data <- 
      gprofiler2_clusters$result %>% 
      mutate(p_adjust = p.adjust(p_value, "bonferroni")) %>% 
      filter(p_adjust < 0.05) %>% 
      filter(term_size < 1000) %>% 
      filter(!source %in% c("TF")) %>% 
      arrange(dplyr::desc(p_adjust)) %>% 
      tail(25) %>% 
      mutate(term_name = factor(term_name, levels = unique(term_name)))
    
    gprofiler2_plot <- 
      ggplot(data    = gprofiler2_plot_data, 
             mapping = aes(x = -log10(p_adjust), y = term_name, fill = -log10(p_adjust), label = paste0(intersection_size,"/", term_size))) + 
      geom_col(show.legend = F) + 
      geom_text(mapping = aes(x = -log10(p_adjust) + .2), hjust = 0) +
      facet_grid(source ~ ., scales = "free", space = "free") + 
      theme_publication() + 
      scale_fill_distiller(palette = "Reds", direction = 1) + 
      theme(strip.text.y       = element_text(angle = 0, hjust = 0), 
            panel.grid.minor.y = element_blank(), 
            panel.grid.major.y = element_blank(), 
            axis.title.y       = element_blank()) +
      scale_x_continuous(expand = expansion(mult = c(0, .2))) +
      labs(x        = "-Log10(adjusted p-value)", 
           fill     = "-Log10(adjusted p-value)", 
           title    = "g:Profiler functional enrichment", 
           subtitle = paste0("Downregulated genes in ", 
                             str_remove(y, ".vs.WT_.*"), 
                             "\n", 
                             "Query size = ", 
                             unique(gprofiler2_plot_data$query_size), 
                             " genes"))
    
    return(gprofiler2_plot)
  })  

future::plan("sequential")


#################################
# enrichGO
#################################
future::plan("multisession", workers = 8)

# enrichGO upregulated 
enrichGO_clusters_upregulated <- 
furrr::future_map(c("4Fj", '4FsB', "4Fk", "4FsA") %>% set_names(., paste(., "upregulated")), function(x){
  
  DGE_results_list <- 
    DGE_genotype_WT_tissue_results[names(DGE_genotype_WT_tissue_results) %>% .[str_detect(., x)]] %>% 
    map(~ .x$geneID_upregulated)
  
  names(DGE_results_list) <- 
    paste(str_remove(names(DGE_results_list), ".*.vs.WT_") %>% str_replace_all("_", " "),
          x,
          "upregulated")
  
  compareCluster(geneClusters  = DGE_results_list,
                 fun           = "enrichGO", 
                 OrgDb         = "org.Mm.eg.db",
                 keyType       = 'ENSEMBL',
                 ont           = "BP",
                 pAdjustMethod = "BH",
                 qvalueCutoff  = 0.05,
                 pvalueCutoff  = 0.05,
                 readable      = TRUE)
  
})  
  
# enrichGO downregulated 
enrichGO_clusters_downregulated <- 
furrr::future_map(c("4Fj", '4FsB', "4Fk", "4FsA") %>% set_names(., paste(., "downregulated")), function(x){
  
  DGE_results_list <- 
    DGE_genotype_WT_tissue_results[names(DGE_genotype_WT_tissue_results) %>% .[str_detect(., x)]] %>% 
    map(~ .x$geneID_downregulated)
  
  names(DGE_results_list) <- 
    paste(str_remove(names(DGE_results_list), ".*.vs.WT_") %>% str_replace_all("_", " "),
          x,
          "downregulated")
  
  compareCluster(geneClusters  = DGE_results_list,
                 fun           = "enrichGO", 
                 OrgDb         = "org.Mm.eg.db",
                 keyType       = 'ENSEMBL',
                 ont           = "BP",
                 pAdjustMethod = "BH",
                 qvalueCutoff  = 0.05,
                 pvalueCutoff  = 0.05,
                 readable      = TRUE)
  
})

future::plan("sequential")

# enrichGO dotplot
enrichGO_clusters_dotplot <-
  map(c(enrichGO_clusters_upregulated, enrichGO_clusters_downregulated), function(x){
    
    clusterProfiler::dotplot(x, group = F, showCategory = 4) +
      scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 50), drop = FALSE) +
      labs(color = "Adj. p-value", size = "Gene ratio") +
      theme(panel.grid.major = element_line(linetype = 2), 
            axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
            axis.title.x     = element_blank()) +
      scale_color_distiller(palette = "Reds", direction = -1)
    
  })


#################################
# emapplot
#################################
enrichGO_clusters_emapplot <- 
map2(c(enrichGO_clusters_upregulated, enrichGO_clusters_downregulated), names(c(enrichGO_clusters_upregulated, enrichGO_clusters_downregulated)), 
    
  ~ emapplot(enrichplot::pairwise_termsim(.x), showCategory = 3) + 
    scale_fill_manual(values = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D") %>%
                               set_names(paste(levels(SummarizedExperiment_object$Tissue), .y)))
  
)


#################################
# enrichKEGG
#################################
ensembl_id_to_entrez_id <- 
  function(gene_id_vector){
    
    mapIds(x         = org.Mm.eg.db, 
           keys      = gene_id_vector,
           column    = "ENTREZID",
           keytype   = "ENSEMBL",
           multiVals = "first") %>% 
      na.omit
    
  }

# enrichKEGG upregulated 
enrichKEGG_clusters_upregulated <- 
map(c("4Fj", '4FsB', "4Fk", "4FsA") %>% set_names(., paste(., "upregulated")), function(x){
  
  DGE_results_list <- 
    DGE_genotype_WT_tissue_results[names(DGE_genotype_WT_tissue_results) %>% .[str_detect(., x)]] %>% 
    map(~ ensembl_id_to_entrez_id(.x$geneID_upregulated))
  
  names(DGE_results_list) <- 
    paste(str_remove(names(DGE_results_list), ".*.vs.WT_") %>% str_replace_all("_", " "),
          x,
          "upregulated")
  
  compareCluster(geneClusters  = DGE_results_list,
                 fun           = "enrichKEGG", 
                 organism      = "mmu",
                 pAdjustMethod = "BH",
                 qvalueCutoff  = 0.05,
                 pvalueCutoff  = 0.05)
  
})  
  
# enrichKEGG downregulated 
enrichKEGG_clusters_downregulated <- 
map(c("4Fj", '4FsB', "4Fk", "4FsA") %>% set_names(., paste(., "downregulated")), function(x){
  
  DGE_results_list <- 
    DGE_genotype_WT_tissue_results[names(DGE_genotype_WT_tissue_results) %>% .[str_detect(., x)]] %>% 
    map(~ ensembl_id_to_entrez_id(.x$geneID_downregulated))
  
  names(DGE_results_list) <- 
    paste(str_remove(names(DGE_results_list), ".*.vs.WT_") %>% str_replace_all("_", " "),
          x,
          "downregulated")
  
  compareCluster(geneClusters  = DGE_results_list,
                 fun           = "enrichKEGG", 
                 organism      = "mmu",
                 pAdjustMethod = "BH",
                 qvalueCutoff  = 0.05,
                 pvalueCutoff  = 0.05)
  
})

# enrichKEGG dotplot
enrichKEGG_clusters_dotplot <-
  map(c(enrichKEGG_clusters_upregulated, enrichKEGG_clusters_downregulated), function(x){
    
    x@compareClusterResult$Description <- str_remove(x@compareClusterResult$Description, " \\- Mus musculus \\(house mouse\\)")
    
    clusterProfiler::dotplot(x, group = F, showCategory = 4) +
      scale_y_discrete(labels = function(x) str_wrap(str_to_sentence(x), width = 50), drop = FALSE) +
      labs(color = "Adj. p-value", size = "Gene ratio") +
      theme(panel.grid.major = element_line(linetype = 2), 
            axis.text.x      = element_text(hjust = 1, vjust = 1, angle = 45), 
            axis.title.x     = element_blank()) +
      scale_color_distiller(palette = "Reds", direction = -1)
    
  })


#################################
# Volcano plot
#################################

volcano_plots <- 
map2(DGE_results$DGE_genotype_WT_tissue, names(DGE_results$DGE_genotype_WT_tissue), function(x, y){
  
  DGE_results_w_gene_anno <- 
      x %>% 
      as.data.frame() %>% 
      rownames_to_column("gene_id") %>% 
      left_join(as.data.frame(rowData(SummarizedExperiment_object)))
  
  geom_label_data <- 
    rbind(
      DGE_results_w_gene_anno %>% filter(abs(log2FoldChange) > 2 & padj < 0.05) %>% arrange(padj) %>% head(10),
      DGE_results_w_gene_anno %>% filter(abs(log2FoldChange) > 2 & padj < 0.05) %>% arrange(dplyr::desc(abs(log2FoldChange))) %>% head(15)
    ) %>% 
    unique
  
  non_signficant_data <- 
    DGE_results_w_gene_anno %>% 
    filter(abs(log2FoldChange) <= 2 | padj >= 0.05)
  
  signficant_data <- 
    DGE_results_w_gene_anno %>% 
    filter(abs(log2FoldChange) > 2 & padj < 0.05)
  
  ggplot(data = DGE_results_w_gene_anno, aes(log2FoldChange, -log10(padj), color = -log10(padj))) + 
    geom_hline(yintercept = 0) +
    geom_vline(xintercept = 0) +
    geom_hline(yintercept = -log10(0.05), color = "darkgray", lty = 2) +
    geom_vline(xintercept = 2, color = "darkgray", lty = 2) +
    geom_vline(xintercept = -2, color = "darkgray", lty = 2) +
    geom_point(data = non_signficant_data, color = "gray") +
    geom_point(data = signficant_data, size = 3) +
    ggrepel::geom_label_repel(data = geom_label_data, mapping = aes(label = gene_name), color = "black", box.padding = unit(0.1, "in")) +
    theme_publication() +
    scale_x_continuous(limits = ggpmisc::symmetric_limits) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), 
          legend.direction = "vertical", 
          legend.position  = "right", 
          plot.title       = element_text(face = "bold")) +
    labs(x        = "Log2 Fold change", 
         y        = "-Log10 adj. P-value", 
         color    = "-Log10 adj. P-value", 
         subtitle = paste0("Differential gene expression between ", str_remove(y, ".vs.WT_.*"), " and WT"), 
         title    = str_remove(y, ".*.vs.WT_") %>% str_replace_all("_", " ")) +
    paletteer::scale_color_paletteer_c("ggthemes::Orange-Blue Diverging", direction = -1)

})

```

# Save plots
```{r}

svg(filename = "output/figures/05_DGE_genotype_WT_tissue/plot_DGE_number_pct.svg", width = 15, height = 10)
plot_DGE_number / plot_DGE_pct
dev.off()

```

# Session information
```{r Session info, echo = FALSE, results = "markup"}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.
