---
title:    "Ocampo Lab mouse reprogramming"
subtitle: "Differential expressed genes"
author:   "Author: Jo√£o Agostinho de Sousa"
date:     "Date: `r format(Sys.time(),'%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    theme: default
    code_folding: hide
editor_options:
  chunk_output_type: console
---

# Load libraries
```{r}

library(tidyverse)
library(DESeq2)

```

# Load objects
```{r}

# Metadata
load(file = "output/database/metadata.RData")

# tximport object
load(file = "output/database/tximport_object.RData")

```

# Differential expression between conditions and WT independent of the tissue
```{r}

# Subset metadata to match the tximport object samples
metadata_subset <- 
  metadata %>% 
  filter(`Sample ID` %in% colnames(tximport_object$counts))
  
# Check if the sample IDs in the metadata are the same and in the same order as in the tximport object
stopifnot(identical(as.character(metadata_subset$`Sample ID`), colnames(tximport_object$counts)))
  
# DESeq2 dataset
deseq2_object <- 
  DESeq2::DESeqDataSetFromTximport(txi     = tximport_object, 
                                   colData = metadata_subset, 
                                   design  = ~ Tissue + Genotype)

DGE_genotype_WT <- 
map(list("4Fj.vs.WT"  = c("4Fj",  "WT"),
         "4FsB.vs.WT" = c("4FsB", "WT"),
         "4Fk.vs.WT"  = c("4Fk",  "WT"),
         "4FsA.vs.WT" = c("4FsA", "WT")
         ),
    function(x){

  # Subset samples to match the genotypes being tested
  sample_subset <- 
    metadata_subset %>% 
    filter(Genotype %in% x) %>% 
    pull(`Sample ID`) %>% 
    as.character()
  
  # Subset the DESeq2 object and create new Genotype levels
  deseq2_object <- deseq2_object[,colnames(deseq2_object) %in% sample_subset]
  deseq2_object$Genotype <- factor(deseq2_object$Genotype, levels = x[2:1])
  
  deseq2 <- DESeq(object = deseq2_object, parallel = T, modelMatrixType = "standard", BPPARAM = BiocParallel::MulticoreParam(20))
    
  deseq2_results <- DESeq2::results(deseq2, alpha = 0.05, parallel = T, BPPARAM = BiocParallel::MulticoreParam(20))
  
  return(deseq2_results)

})

```

# Differential expression between conditions and WT for each tissue
```{r}

# Subset metadata to match the tximport object samples
metadata_subset <- 
  metadata %>% 
  filter(`Sample ID` %in% colnames(tximport_object$counts))
  
# Check if the sample IDs in the metadata are the same and in the same order as in the tximport object
stopifnot(identical(as.character(metadata_subset$`Sample ID`), colnames(tximport_object$counts)))
  
# DESeq2 dataset
deseq2_object <- 
  DESeq2::DESeqDataSetFromTximport(txi     = tximport_object, 
                                   colData = metadata_subset, 
                                   design  = ~ Genotype)

DGE_genotype_WT_tissue <- 
map(list("4Fj.vs.WT_Kidney"          = c("4Fj",  "WT", "Kidney"),
         "4Fj.vs.WT_Liver"           = c("4Fj",  "WT", "Liver"),
         "4Fj.vs.WT_Small_intestine" = c("4Fj",  "WT", "Small intestine"),
         "4Fj.vs.WT_Heart"           = c("4Fj",  "WT", "Heart"),
         "4Fj.vs.WT_Brain"           = c("4Fj",  "WT", "Brain"),
         "4Fj.vs.WT_Skeletal_muscle" = c("4Fj",  "WT", "Skeletal muscle"),
         "4Fj.vs.WT_Spleen"          = c("4Fj",  "WT", "Spleen"),

         "4FsB.vs.WT_Kidney"          = c("4FsB",  "WT", "Kidney"),
         "4FsB.vs.WT_Liver"           = c("4FsB",  "WT", "Liver"),
         "4FsB.vs.WT_Small_intestine" = c("4FsB",  "WT", "Small intestine"),
         "4FsB.vs.WT_Heart"           = c("4FsB",  "WT", "Heart"),
         "4FsB.vs.WT_Brain"           = c("4FsB",  "WT", "Brain"),
         "4FsB.vs.WT_Skeletal_muscle" = c("4FsB",  "WT", "Skeletal muscle"),

         "4Fk.vs.WT_Kidney"          = c("4Fk",  "WT", "Kidney"),
         "4Fk.vs.WT_Liver"           = c("4Fk",  "WT", "Liver"),
         "4Fk.vs.WT_Small_intestine" = c("4Fk",  "WT", "Small intestine"),
         "4Fk.vs.WT_Heart"           = c("4Fk",  "WT", "Heart"),
         "4Fk.vs.WT_Brain"           = c("4Fk",  "WT", "Brain"),
         "4Fk.vs.WT_Skeletal_muscle" = c("4Fk",  "WT", "Skeletal muscle"),

         "4FsA.vs.WT_Kidney"          = c("4FsA",  "WT", "Kidney"),
         "4FsA.vs.WT_Liver"           = c("4FsA",  "WT", "Liver"),
         "4FsA.vs.WT_Small_intestine" = c("4FsA",  "WT", "Small intestine"),
         "4FsA.vs.WT_Heart"           = c("4FsA",  "WT", "Heart"),
         "4FsA.vs.WT_Brain"           = c("4FsA",  "WT", "Brain"),
         "4FsA.vs.WT_Skeletal_muscle" = c("4FsA",  "WT", "Skeletal muscle"),
         "4FsA.vs.WT_Spleen"          = c("4FsA",  "WT", "Spleen")
         ),
    function(x){

  # Subset samples to match the genotypes being tested
  sample_subset <- 
    metadata_subset %>% 
    filter(Tissue == x[3]) %>% 
    filter(Genotype %in% x[1:2]) %>% 
    pull(`Sample ID`)
  
  # Subset the DESeq2 object and create new Genotype levels
  deseq2_object <- deseq2_object[,colnames(deseq2_object) %in% sample_subset]
  deseq2_object$Genotype <- factor(deseq2_object$Genotype, levels = x[2:1])
  
  deseq2 <- DESeq(object = deseq2_object, parallel = T, modelMatrixType = "standard", BPPARAM = BiocParallel::MulticoreParam(20))
    
  deseq2_results <- DESeq2::results(deseq2, alpha = 0.05, parallel = T, BPPARAM = BiocParallel::MulticoreParam(20))
  
  return(deseq2_results)

})

```

# Differential expression between conditions for each tissue
```{r}

# Subset metadata to match the tximport object samples
metadata_subset <- 
  metadata %>% 
  filter(`Sample ID` %in% colnames(tximport_object$counts))
  
# Check if the sample IDs in the metadata are the same and in the same order as in the tximport object
stopifnot(identical(as.character(metadata_subset$`Sample ID`), colnames(tximport_object$counts)))
  
# DESeq2 dataset
deseq2_object <- 
  DESeq2::DESeqDataSetFromTximport(txi     = tximport_object, 
                                   colData = metadata_subset, 
                                   design  = ~ Genotype)

DGE_genotype_tissue <- 
map(list(
         "4Fj.vs.4FsB_Kidney"  = c("4Fj",  "4FsB", "Kidney"),
         "4Fj.vs.4Fk_Kidney"   = c("4Fj",  "4Fk",  "Kidney"),
         "4Fj.vs.4FsA_Kidney"  = c("4Fj",  "4FsA", "Kidney"),
         "4FsB.vs.4Fk_Kidney"  = c("4FsB", "4Fk",  "Kidney"),
         "4FsB.vs.4FsA_Kidney" = c("4FsB", "4FsA", "Kidney"),
         "4Fk.vs.4FsA_Kidney"  = c("4Fk",  "4FsA", "Kidney"),

         "4Fj.vs.4FsB_Liver"  = c("4Fj",  "4FsB", "Liver"),
         "4Fj.vs.4Fk_Liver"   = c("4Fj",  "4Fk",  "Liver"),
         "4Fj.vs.4FsA_Liver"  = c("4Fj",  "4FsA", "Liver"),
         "4FsB.vs.4Fk_Liver"  = c("4FsB", "4Fk",  "Liver"),
         "4FsB.vs.4FsA_Liver" = c("4FsB", "4FsA", "Liver"),
         "4Fk.vs.4FsA_Liver"  = c("4Fk",  "4FsA", "Liver"),

         "4Fj.vs.4FsB_Small_intestine"  = c("4Fj",  "4FsB", "Small intestine"),
         "4Fj.vs.4Fk_Small_intestine"   = c("4Fj",  "4Fk",  "Small intestine"),
         "4Fj.vs.4FsA_Small_intestine"  = c("4Fj",  "4FsA", "Small intestine"),
         "4FsB.vs.4Fk_Small_intestine"  = c("4FsB", "4Fk",  "Small intestine"),
         "4FsB.vs.4FsA_Small_intestine" = c("4FsB", "4FsA", "Small intestine"),
         "4Fk.vs.4FsA_Small_intestine"  = c("4Fk",  "4FsA", "Small intestine"),

         "4Fj.vs.4FsB_Heart"  = c("4Fj",  "4FsB", "Heart"),
         "4Fj.vs.4Fk_Heart"   = c("4Fj",  "4Fk",  "Heart"),
         "4Fj.vs.4FsA_Heart"  = c("4Fj",  "4FsA", "Heart"),
         "4FsB.vs.4Fk_Heart"  = c("4FsB", "4Fk",  "Heart"),
         "4FsB.vs.4FsA_Heart" = c("4FsB", "4FsA", "Heart"),
         "4Fk.vs.4FsA_Heart"  = c("4Fk",  "4FsA", "Heart"),

         "4Fj.vs.4FsB_Brain"  = c("4Fj",  "4FsB", "Brain"),
         "4Fj.vs.4Fk_Brain"   = c("4Fj",  "4Fk",  "Brain"),
         "4Fj.vs.4FsA_Brain"  = c("4Fj",  "4FsA", "Brain"),
         "4FsB.vs.4Fk_Brain"  = c("4FsB", "4Fk",  "Brain"),
         "4FsB.vs.4FsA_Brain" = c("4FsB", "4FsA", "Brain"),
         "4Fk.vs.4FsA_Brain"  = c("4Fk",  "4FsA", "Brain"),

         "4Fj.vs.4FsB_Skeletal_muscle"  = c("4Fj",  "4FsB", "Skeletal muscle"),
         "4Fj.vs.4Fk_Skeletal_muscle"   = c("4Fj",  "4Fk",  "Skeletal muscle"),
         "4Fj.vs.4FsA_Skeletal_muscle"  = c("4Fj",  "4FsA", "Skeletal muscle"),
         "4FsB.vs.4Fk_Skeletal_muscle"  = c("4FsB", "4Fk",  "Skeletal muscle"),
         "4FsB.vs.4FsA_Skeletal_muscle" = c("4FsB", "4FsA", "Skeletal muscle"),
         "4Fk.vs.4FsA_Skeletal_muscle"  = c("4Fk",  "4FsA", "Skeletal muscle"),

         "4Fj.vs.4FsA_Spleen"  = c("4Fj",  "4FsA", "Spleen")
         ),
    function(x){

  # Subset samples to match the genotypes being tested
  sample_subset <- 
    metadata_subset %>% 
    filter(Tissue == x[3]) %>% 
    filter(Genotype %in% x[1:2]) %>% 
    pull(`Sample ID`)
  
  # Subset the DESeq2 object and create new Genotype levels
  deseq2_object <- deseq2_object[,colnames(deseq2_object) %in% sample_subset]
  deseq2_object$Genotype <- factor(deseq2_object$Genotype, levels = x[2:1])
  
  deseq2 <- DESeq(object = deseq2_object, parallel = T, modelMatrixType = "standard", BPPARAM = BiocParallel::MulticoreParam(20))
    
  deseq2_results <- DESeq2::results(deseq2, alpha = 0.05, parallel = T, BPPARAM = BiocParallel::MulticoreParam(20))
  
  return(deseq2_results)

})

```

# Differential expression between one tissue and the remaining tissue per genotype
```{r}

# Subset metadata to match the tximport object samples
metadata_subset <- 
  metadata %>% 
  filter(`Sample ID` %in% colnames(tximport_object$counts))
  
# Check if the sample IDs in the metadata are the same and in the same order as in the tximport object
stopifnot(identical(as.character(metadata_subset$`Sample ID`), colnames(tximport_object$counts)))
  
# DESeq2 dataset
deseq2_object <- 
  DESeq2::DESeqDataSetFromTximport(txi     = tximport_object, 
                                   colData = metadata_subset, 
                                   design  = ~ Tissue)

DGE_tissue <- 
map(list(
  
  "WT_Kidney.vs.WT_other_tissues"          = c("WT", "Kidney"),
  "WT_Liver.vs.WT_other_tissues"           = c("WT", "Liver"),
  "WT_Small_intestine.vs.WT_other_tissues" = c("WT", "Small intestine"),
  "WT_Heart.vs.WT_other_tissues"           = c("WT", "Heart"),
  "WT_Brain.vs.WT_other_tissues"           = c("WT", "Brain"),
  "WT_Skeletal_muscle.vs.WT_other_tissues" = c("WT", "Skeletal muscle"),
  "WT_Spleen.vs.WT_other_tissues"          = c("WT", "Spleen"),
  
  "4Fj_Kidney.vs.4Fj_other_tissues"          = c("4Fj", "Kidney"),
  "4Fj_Liver.vs.4Fj_other_tissues"           = c("4Fj", "Liver"),
  "4Fj_Small_intestine.vs.4Fj_other_tissues" = c("4Fj", "Small intestine"),
  "4Fj_Heart.vs.4Fj_other_tissues"           = c("4Fj", "Heart"),
  "4Fj_Brain.vs.4Fj_other_tissues"           = c("4Fj", "Brain"),
  "4Fj_Skeletal_muscle.vs.4Fj_other_tissues" = c("4Fj", "Skeletal muscle"),
  "4Fj_Spleen.vs.4Fj_other_tissues"          = c("4Fj", "Spleen"),
  
  "4FsB_Kidney.vs.4FsB_other_tissues"          = c("4FsB", "Kidney"),
  "4FsB_Liver.vs.4FsB_other_tissues"           = c("4FsB", "Liver"),
  "4FsB_Small_intestine.vs.4FsB_other_tissues" = c("4FsB", "Small intestine"),
  "4FsB_Heart.vs.4FsB_other_tissues"           = c("4FsB", "Heart"),
  "4FsB_Brain.vs.4FsB_other_tissues"           = c("4FsB", "Brain"),
  "4FsB_Skeletal_muscle.vs.4FsB_other_tissues" = c("4FsB", "Skeletal muscle"),
  
  "4Fk_Kidney.vs.4Fk_other_tissues"          = c("4Fk", "Kidney"),
  "4Fk_Liver.vs.4Fk_other_tissues"           = c("4Fk", "Liver"),
  "4Fk_Small_intestine.vs.4Fk_other_tissues" = c("4Fk", "Small intestine"),
  "4Fk_Heart.vs.4Fk_other_tissues"           = c("4Fk", "Heart"),
  "4Fk_Brain.vs.4Fk_other_tissues"           = c("4Fk", "Brain"),
  "4Fk_Skeletal_muscle.vs.4Fk_other_tissues" = c("4Fk", "Skeletal muscle"),
  
  "4FsA_Kidney.vs.4FsA_other_tissues"          = c("4FsA", "Kidney"),
  "4FsA_Liver.vs.4FsA_other_tissues"           = c("4FsA", "Liver"),
  "4FsA_Small_intestine.vs.4FsA_other_tissues" = c("4FsA", "Small intestine"),
  "4FsA_Heart.vs.4FsA_other_tissues"           = c("4FsA", "Heart"),
  "4FsA_Brain.vs.4FsA_other_tissues"           = c("4FsA", "Brain"),
  "4FsA_Skeletal_muscle.vs.4FsA_other_tissues" = c("4FsA", "Skeletal muscle"),
  "4FsA_Spleen.vs.4FsA_other_tissues"          = c("4FsA", "Spleen")
         ),
    function(x){

  # Subset samples to match the genotypes being tested
  sample_subset <- 
    metadata_subset %>% 
    filter(Genotype %in% x[1]) %>% 
    mutate(Tissue = case_when(Tissue == x[2] ~ Tissue, 
                              TRUE ~ "Other tissues")) %>%
    pull(`Sample ID`)
  
  # Subset the DESeq2 object and create new Genotype levels
  deseq2_object <- deseq2_object[,colnames(deseq2_object) %in% sample_subset]
  deseq2_object$Tissue <- factor(deseq2_object$Tissue, levels = c("Other tissues", x[2]))
  deseq2_object$Tissue <- deseq2_object$Tissue %>% str_replace_na("Other tissues")
  deseq2_object$Tissue <- factor(deseq2_object$Tissue, levels = c("Other tissues", x[2]))
  
  deseq2 <- DESeq(object = deseq2_object, parallel = T, modelMatrixType = "standard", BPPARAM = BiocParallel::MulticoreParam(20))
    
  deseq2_results <- DESeq2::results(deseq2, alpha = 0.05, parallel = T, BPPARAM = BiocParallel::MulticoreParam(20))
  
  return(deseq2_results)

})

```

# Save object
```{r}

DGE_results <- 
list(
"DGE_genotype_WT"        = DGE_genotype_WT,
"DGE_genotype_WT_tissue" = DGE_genotype_WT_tissue,
"DGE_genotype_tissue"    = DGE_genotype_tissue,
"DGE_tissue"             = DGE_tissue
)

save(DGE_results, file = "output/database/DGE_results.RData")

```

# Session information
```{r Session info, echo = FALSE, results = "markup"}

devtools::session_info()

```

This document was processed on: `r Sys.Date()`.
